### PR TITLE
TSL: Introduce `renderOutput()`

### PR DESCRIPTION
**Pull request by sunag**
Related issue: https://github.com/mrdoob/three.js/pull/28779#discussion_r1661371743

**Description**

Color Space Conversion and Tone Mapping are applied in Post Processing pass, it is possible to manipulate the sequence too using `renderOutput()` as updated in the `3dlut` example.

```js
// post processing

postProcessing = new THREE.PostProcessing( renderer );

// ignore default output color transform ( toneMapping and outputColorSpace )
// use renderOutput() for control the sequence

postProcessing.outputColorTransform = false;

// scene pass

const scenePass = pass( scene, camera );
const outputPass = renderOutput( scenePass );

lutPass = outputPass.lut3D();
lutPass.lutNode = texture3D( lutMap[ params.lut ] );
lutPass.intensityNode = uniform( 1 );

postProcessing.outputNode = lutPass;
```

`.toneMappingNode` has been removed and the examples updated, I think it can be replaced by post-processing now.

### CODE REVIEW

**Comment by sunag** [comment]
https://github.com/mrdoob/three.js/pull/28780 It seems to work :)

File: src/renderers/common/PostProcessing.js [hunk 1]
```diff
@@ -10,27 +11,73 @@ class PostProcessing {
 		this.renderer = renderer;
 		this.outputNode = outputNode;
 
+		this.defaultColorTransform = true;
```
Comment (donmccurdy) [inline src/renderers/common/PostProcessing.js:14]: There may be other color transforms involved in a post-processing stack, and I doubt we want this flag to disable all of them — could I recommend the word "output" be part of this name to clarify? Perhaps `outputColorTransform`?

A common, related term in color management is Output Device Transform (ODT). Though, I'm hesitant to use that exactly here, as we're mapping to a canvas drawing buffer's color space, or perhaps a framebuffer, and not necessarily for a known 'device'.

/cc @WestLangley

File: src/renderers/common/nodes/Nodes.js [hunk 2]
```diff
@@ -391,29 +393,30 @@ class Nodes extends DataMap {
...
 
-		}
+	getOutputNode( outputTexture ) {
 
-		if ( this.renderer.currentColorSpace === SRGBColorSpace ) {
+		const renderer = this.renderer;
+		const cacheKey = this.getOutputCacheKey();
 
-			output = output.linearToColorSpace( this.renderer.currentColorSpace );
+		const output = texture( outputTexture, viewportTopLeft ).renderOutput( renderer.toneMapping, renderer.currentColorSpace );
```
Comment (donmccurdy) [inline src/renderers/common/nodes/Nodes.js:417]: What does `renderer.currentColorSpace` represent? What is "current" in this context?

Comment (sunag) [reply src/renderers/common/nodes/Nodes.js:417]: `renderer.currentColorSpace` considers the color space used by the `RenderTarget` if it is defined.

**Review by Mugen87 — APPROVED** [review]
Awesome! This change solves both issues discussed in #28779. Devs can now controls via `RenderOutputNode` when output tone mapping and color space conversion should happen in the pass chain. And second, the additional output pass in `Renderer` is now disabled when post processing is used (so there is on separate output pass anymore).

File: src/renderers/common/PostProcessing.js (hunk 1 above)
Comment (Mugen87) [reply src/renderers/common/PostProcessing.js:14]: I think what the property wants to express is that there is a default color transformation in `PostProcessing` which the developer can disable when `RenderOutputPass` is used. So maybe `defaultOutputColorTransform`?

Comment (sunag) [reply src/renderers/common/PostProcessing.js]: > What the property wants to express is that there is a default color transformation in PostProcessing which the developer can disable when this step is done manually

I had this doubt too, but I believe that `outputColorTransform` can convey this idea too.

**Comment by Mugen87** [comment]
Unlike `Lut3DNode`, some passes like FXAA require the result of `RenderOutputNode` as a texture. But if you call `toTexture()` on the node, the result is getting to dark. Check out the following fiddle to see the effect:

https://jsfiddle.net/3vgz4sxw/

If you remove the `.toTexture()` call, colors are as expected.

**Comment by WestLangley** [comment]
FWIW, maybe
```js
postProcessing.outputColorTransform = true;
```

**Comment by Mugen87** [comment]
I'm afraid that is not right. The idea is to set `outputColorTransform` to `false` if you use `RenderOutputNode`.

**Comment by donmccurdy** [comment]
I don't think I quite understand this code:

```javascript
  // post processing

  postProcessing = new THREE.PostProcessing( renderer );
  postProcessing.outputColorTransform = false;

  // scene pass

  const scenePass = pass( scene, camera );
  const outputPass = renderOutput( scenePass );


  postProcessing.outputNode = outputPass.toTexture();
```

Correct me if I get this wrong, but it looks like `scenePass` is the initial render pass, `renderOutput` tone maps and converts to the display color space, and then the result is assigned to `postProcessing.outputNode` ... so `postProcessing.outputNode` is the _input_ to the post-processing stack?

**Comment by Mugen87** [comment]
The final node is assigned to `postProcessing.outputNode`. With `FXAANode`, a typical setup could look like this:
```js
 // post processing

postProcessing = new THREE.PostProcessing( renderer );
postProcessing.outputColorTransform = false;

  // scene pass (ideally produced via MRT)

const scenePass = pass( scene, camera );
const scenePassColor = scenePass.getTextureNode();
const scenePassViewZ = scenePass.getViewZNode();
cosnt scenePassNormal = scenePass.getNormal();

// bloom -> output (tone-maping+sRGB) -> fxaa

const bloomPass = scenePassColor.bloom( scenePassViewZ, scenePassNormal );
const outputPass = bloomPass. renderOutput();
const fxaaPass = outputPass.fxaa();

postProcessing.outputNode =fxaaPass;
```
FXAA requires sRGB input and since it is not a per-pixel operation, the result of previous pass as a texture. The `toTexture()` call is done internally in `FXAANode`. The code in the fiddle is just a simplified setup.

**Comment by Mugen87** [comment]
You can actually chain all commands or write them more explictly:
```js
const bloomPass = scenePassColor.bloom( scenePassViewZ, scenePassNormal );
const outputPass = bloomPass.renderOutput();
const fxaaPass = outputPass.fxaa();

**Comment by sunag** [comment]
@Mugen87 Try add this for now:

```js
const outputPass = renderOutput( scenePass, renderer.toneMapping, renderer.outputColorSpace )
```

`.toTexture()` It is not transferring the context of rendering, but this can be fixed.

**Pull request merged**